RFD 219 - Kubernetes RBAC simplification (#56214)
gravitational/teleport: rfd/0219-k8s-rbac-simplification.md

RBAC (Role-Based Access Control) within the Teleport project.

The goal is to streamline the user experience and reduce the complexity of
getting up and running on day 1 and managing permissions later on.

## Why

Today, the Kubernetes RBAC in Teleport is complex and can be difficult for
users to manage.
A proper setup requires some external Kubernetes config and matching role
setup.
This complexity leads to misconfigurations and security issues.

Simplifying the RBAC model will help improve usability and security.

### References

A common issue currently is the un-intuitive result of complex rule sets and
more critically, difficult to get setup on day 1.

- @klizhentas struggling to setup a cluster following the `self-hosted` flow:
  (internal) <https://gravitational.slack.com/archives/C03SULRAAG3/p1715394587181739>
- Customer confusion on role resolution:
  - <https://github.com/gravitational/teleport/issues/45475>
  - (internal) <https://goteleport.slack.com/archives/C0582MBSMHN/p1738187497666819>
- Unexpected `exec` grant on read-only user:
  (internal) <https://gravitational.slack.com/archives/C32M8FP1V/p1739462454321459?thread_ts=1739462384.714419&cid=C32M8FP1V>

## Goals

> - Discourage the use of `kubernetes_groups` and `kubernetes_users` fields from

Reviewer: you mean discourage for getting started, not in general?

Currently, when enrolling a Kubernetes cluster and setting it up to use the
group `system:masters`, everything works as expected (except on GKE Autopilot
where this group is forbidden).
Teleport is setup to impersonate to yield reduced permissions to the user
based on the `kubernetes_resources` and `kubernetes_label` fields from the
Teleport role.

### Prior work

While initially planned to be included in this RFD, due to customer requests,
the following has already been implemented and shipped:

- Support arbitrary resources, including CRDs #54426
- Simplify Namespace/Kind behavior #54938
- Handle all known resources on the Teleport side #55099

Given this, as RBAC has already been simplified quite a lot, a scoped down
version of the proposal would be to focus on new installs without impacting
existing customers.

### Authoritative RBAC Documentation

To simplify initial setup and long-term management, we'll update the
documentation to encourage users to use a cluster-wide admin group that will be
created by provisioning scripts or auto-discovery.
This will allow users to avoid the complexity of managing RBAC in multiple
places, avoid confusion around the `exec` subresource.

> While we won't change or remove the `kubernetes_groups` and `kubernetes_users`

Reviewer: what should they use instead for fine-grained access?

  - User provisions the Kubernetes cluster using a script, discovery, or Helm
    chart to create the impersonation ClusterRole/ClusterRoleBinding.
  - User configures Kubernetes RBAC with a custom ClusterRole with some
    permissions
  - User configures Kubernetes RBAC with a custom ClusterRoleBinding, which
    needs to be understood matches with the `kubernetes_groups` field in the
    Teleport role.
  - User configures Teleport role with `kubernetes_groups`, (unclear what
    `kubernetes_users` does from the docs) fields to match the custom
    ClusterRoleBinding.
  - User configures Teleport role with `kubernetes_resources` and
    `kubernetes_labels` to match to match or reduce the permissions granted by
    the custom ClusterRole.
- Day 1 - Ongoing management:
  - User needs to understand the interaction between the Teleport role and the
    underlying Kubernetes RBAC.
  - User may need to update the Teleport role if the underlying Kubernetes
    RBAC changes.
  - No clear guidance on how to which permission should be set where (Teleport
    or Kubernetes).
  - User may need to troubleshoot unexpected permissions or access issues.

> #### Proposed flow

Author: We create a new preset to make day 0 easier as there is currently no preset roles granting kubernetes access.
The `kubernetes_groups` field is required and allows Teleport itself to have cluster-admin permissions, allowing us to be authoritative and let the user only worry about labels and resources instead.

- Day 0 - Initial setup:
  - User installs Teleport and sets up a Kubernetes cluster.
  - User provisions the Kubernetes cluster using a script, discovery, or Helm
    chart to create all required RBAC resources.
  - User configures the Teleport role with `kubernetes_resources` and
    `kubernetes_labels` to match or reduce the permissions granted by the
    ClusterRole. The `kubernetes_groups` field is pre-populated in the Web UI
    and well documented in the Teleport documentation for the YAML version.
- Day 1 - Ongoing management:

>   - User can manage the Teleport role without needing to understand the

Reviewer: if that's the case, what is the purpose of `kubernetes_groups` in the preset role `kube-editor`?

Author: While it could be removed for new customers, it's purpose it to maintain compatibility with existing customers who may not be willing to apply the new clusterrole/clusterrolebinding to enable the simplification.

- @klizhentas struggling to setup a cluster following the `self-hosted` flow:
  (internal) <https://gravitational.slack.com/archives/C03SULRAAG3/p1715394587181739>
- Customer confusion on role resolution:
  - <https://github.com/gravitational/teleport/issues/45475>
  - (internal) <https://goteleport.slack.com/archives/C0582MBSMHN/p1738187497666819>
- Unexpected `exec` grant on read-only user:
  (internal) <https://gravitational.slack.com/archives/C32M8FP1V/p1739462454321459?thread_ts=1739462384.714419&cid=C32M8FP1V>

## Goals

- Make the Teleport agent authoritative
- Remove the `kubernetes_groups` and `kubernetes_users` fields from role
  - This will require a role model version bump to v9
- Clarify / Improve documentation for the various ways to setup Kubernetes in
  Teleport.

## Proposal

### Background

Currently, when enrolling a Kubernetes cluster and setting it up to use the
group `system:masters`, everything works as expected (except on GKE Autopilot
where this group is forbidden).
Teleport is super admin and yields reduced permissions to the user based on
the `kubernetes_resources` and `kubernetes_label` fields from the Teleport

> role.

Author: Updated. Thanks for the clarification.

## Goals

- Make the Teleport agent authoritative
- Remove the `kubernetes_groups` and `kubernetes_users` fields from role
  - This will require a role model version bump to v9
- Clarify / Improve documentation for the various ways to setup Kubernetes in
  Teleport.

## Prior work

### CRDs

We recently added support for _all_ Kubernetes resources, including CRDs, which
makes resource management easier as we now use the same resource name/group as
Kubernetes.

### Complete resources support

We now register all resources available in Kubernetes instead of restricting
to only a select few. This results in much more consistent permissions
behavior.

While this is a great step, we still fallback to the underlying Kubernetes
cluster when hitting an unknown resource (which happens when a new CRD gets
created), which means we are not yet authoritative.

### Wildcard Kind / Namespace behavior change

Starting with Role v8, the `namespaces` kind only means the resource itself
instead of the resource everything within it.

Role v8 also changed how the wildcard kind works, it now enforces the
`namespace` field, which allows setting permission for a specific cluster-wide

> or namespaced resource easily.

Author: It was just to highlight was has already been done related to the RFD, but indeed, already merged. Removed the prior work section.

---
kind: role
version: v8
metadata:
  name: kube-auditor
spec:
  allow:
    kubernetes_groups:
      - 'teleport:preset:auditor'
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - api_group: '*'
        kind: '*'
        namespace: '*'
        name: '*'
        verbs:
          - '*'
```

### Provisioning

To enable easy setup, provisioning methods will be updated to create the
required new ClusterRoles and ClusterRoleBindings.

#### Helm Chart

The Helm Chart will provide the ability to specify custom names for all
resources, (clusterrole, clusterrolebinding) and for the Group.

Example:

```yaml
roles: kube,app,discovery
authToken: foo
proxyAddr: example.devteleport.com:443
kubeClusterName: myCluster
rbac:
  accessClusterRoleBindingName: teleport:preset:access
  accessGroupName: teleport:preset:access
  editorClusterRoleBindingName: teleport:preset:editor
  editorGroupName: teleport:preset:editor
  auditorClusterRoleBindingName: teleport:preset:auditor
  auditorGroupName: teleport:preset:auditor
```

Note that if a user decides to change the preset names, they will not be able
to use the preset Teleport roles out of the box, as they will need to update

> them to match the new names.

Reviewer 2: > if we fail to create the new teleport:preset resources in Kubernetes?

The chat won't deploy unless you disable the RBAC creation, and in this case I suppose that you want to handle the RBAC on your own and not user what we provide.

If you want you can also create your own rolebindings for our standard group.

> 

Reviewer 2: Adding more details from the discussions with @klizhentas about why we prefer to not do this ourselves.

```suggestion

Kubernetes might introduce new API endpoints that should be allowed or denied.
Vendoring those roles on Teleport would require us to update the list for every
Kubernetes version, and would potentially cause a permission issues if a user
deploys an old Teleport version in a modern Kubernetes cluster. Relying on the
Kubernetes built-in roles for RBAC frees us from the maintenance burden and
prevents security issues caused by version mismatch.

```

> To enable easy setup, provisioning methods will be updated to create the
> required new ClusterRoles and ClusterRoleBindings.

Reviewer 2: We use the kube built-in ClusterRoles, no need to create them:

```suggestion
To enable easy setup, provisioning methods will be updated to create the
required new ClusterRoleBindings.
```

```yaml

{{- if .Values.rbac.defaultClusterRoleBindings -}}
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding

> metadata:metadata:

Reviewer 2: ```suggestion
metadata:
```

  name: {{ .Release.Name }}-{{ .Release.Namespace }}-preset-edit
```

#### Provision Script

For simplicity, the provision script will be updated to always create the
Kubernetes resources, and will not provide a means for users to opt out of

> the new behavior.

Reviewer 2: Not sure which script we are talking about

#### Auto Discovery

##### EKS / AKS

For both EKS and AKS, auto-discovery will be updated to create the expected
RBAC resources.

##### GKE

For GKE, we will need to change the documented GCP IAM role to enable admin
privileges.
Teleport will attempt to impersonate a privileged group, but there is a risk
GKE users will not be able to use the new preset roles out of the box.

A special attention to error management will be required to help users
understand the issue and how to resolve it, which can be easily done by
running the provision script for example.

#### Auto-update

Auto-update will not attempt to apply the new model, with the assumption that
existing setups are already working.

### UX

> #### Exec confusion
> 
> As the underlying Kubernetes RBAC gets discouraged, it will help with the
> confusion around _exec_ being allowed with a `get` verb is gone.
> Teleport uses a dedicated verb to control access to `exec`.
> 

Reviewer 2: ```suggestion
```

#### Resource enrollment UI

On the Web UI, the initial page generates a _Helm_ command line.

After enrollment, a test page is shown, prompting for a `kubernetes_groups`
value, which defaults to the user's trait and updates it if changed in this
page.

The following page is about testing the connection, and also have a form to
populate a subset of the users/groups.

The initial idea was to allow the user to update it's traits on the fly to not
have to logout/log back in, however, as there is no mention of traits, and
because the form is present on multiple pages with different behavior, and
because it requires the user to initially already have a role assigned that
contains the trait interpolation, we will remove both the 'Setup access' page

> and the 'users/groups' form from the 'Test connection' page.

Reviewer 2: We might need to leave them for existing Teleport users who don't have the kube_user and kube_group traits. We can populate them by default though. @marcoandredinis is the owner of the discovery UI process and might have opinions on this.

  - User provisions the Kubernetes cluster using a script, discovery, or Helm
    chart to create the impersonation ClusterRole/ClusterRoleBinding.
  - User configures Kubernetes RBAC with a custom ClusterRole with some
    permissions
  - User configures Kubernetes RBAC with a custom ClusterRoleBinding, which
    needs to be understood matches with the `kubernetes_groups` field in the
    Teleport role.
  - User configures Teleport role with `kubernetes_groups`, (unclear what
    `kubernetes_users` does from the docs) fields to match the custom
    ClusterRoleBinding.
  - User configures Teleport role with `kubernetes_resources` and
    `kubernetes_labels` to match to match or reduce the permissions granted by
    the custom ClusterRole.
- Day 1 - Ongoing management:
  - User needs to understand the interaction between the Teleport role and the
    underlying Kubernetes RBAC.
  - User may need to update the Teleport role if the underlying Kubernetes
    RBAC changes.
  - No clear guidance on how to which permission should be set where (Teleport
    or Kubernetes).
  - User may need to troubleshoot unexpected permissions or access issues.

#### Proposed flow

> - Day 0 - Initial setup:
>   - User installs Teleport and sets up a Kubernetes cluster.
>   - User provisions the Kubernetes cluster using a script, discovery, or Helm
>     chart to create all required RBAC resources.
>   - User configures the Teleport role with `kubernetes_resources` and
>     `kubernetes_labels` to match or reduce the permissions granted by the
>     ClusterRole.
>   - User applies the desired `kubernetes_groups` trait to the Teleport user.
> - Day 1 - Ongoing management:
>   - User can reduce the scope of either Kubernetes or Teleport's roles by
>     changing the counterpart
>   - User has the option to use the `teleport:preset:editor` role to avoid the
>     complexity of managing the underlying Kubernetes RBAC, only managing
>     resources/labels on the Teleport side instead.

Reviewer 2: ```suggestion
- Day 0 - Initial setup, prospect and hackers:
  - Prospect installs Teleport
  - Prospect creates its first Teleport local user with the trait `kubernetes_groups: "teleport:preset:cluster-admin"`
  - Prospect deploys a Teleport agent in their Kubernetes cluster
  - Prospect can access the Kubernetes cluster as administrator with their local Teleport user
- Day 1 - Multiple users, late stage prospect or new Teleport adopter
  - Admin has some Kubernetes agents connected, SSO connector set up, and several users
  - Admin creates per-team roles with different `spec.allow. kubernetes_labels` to select each team's cluster
  - Admin sets the `teleport:preset:edit` or `teleport:preset:view` as `spec.allow.kubernetes_groups`
  - Admin grants the role to users using SSO connector or Access Lists
  - users from each team can access their Kubernetes clusters
- Day 2 - More granular access, advanced Teleport users
  - Admin can create custom Kubernetes rolebindings, deploy them in their clusters and use them in their Teleport roles instead of the presets
  - Admin can scope down read and edit roles to select namespaces or resources by using `kubernmetes_resources` in their roles
```

    networkpolicies, leases

This resulted in the following Teleport roles:

- `editor`, which maps to `cluster-admin` and can do everything.
- `access`, which maps to `edit` and can do most CRUD operations.
- `auditor`, which maps to `view` and can do read-only operations.

#### New Kubernetes Bindings

In order to be consistent and allow for seamless use of Kubernetes within
Teleport, we will introduce the following new preset bindings:

- `kubernetes_groups` -> `ClusterRole`
- `teleport:preset:edit` -> `edit`
  Can CRUD most basic builtin resources and read some limited administrative resources.
- `teleport:preset:cluster-admin` -> `cluster-admin`
  Provides full access to all Kubernetes resources, including CRDs.
- `teleport:preset:view` -> `view`
  Provides read-only access to some Kubernetes resources.

On the Kubernetes side, we will create the corresponding `ClusterRoleBindings` based on `kubernetes_groups`.

While there is an existing binding for `cluster-admin` via `system:masters`, it
is not available everywhere and there are no binding for `edit` nor `view`.
For consistency and reliability, we will create one binding for each preset
role.

> Those bindings will be available to Teleport users via the user trait feature.

Reviewer 3: I see now that it was [discussed above](https://github.com/gravitational/teleport/pull/56214#discussion_r2310361102). TBH I don't see a problem or a consistency issue with including more preset roles if it improves user experience (which I do think it does compared to having to assign traits) - e.g. AWS has literally hundreds of builtin/preset roles for each imaginable scenario which makes it easy to find and pick one you need (or use as a base to tweak policies). In fact, those specialized roles I think would actually be better from security standpoint than relying on blanket access role as well, and if somebody wants to do this via traits, they can too. But like mentioned above I don't think it's a blocker.

mentioning the required steps to setup permissions properly. Deep error
management is already used for GKE where `system:masters` isn't allowed.

To know which group to use to be authoritative (see below), we need a config
value. A workaround can be implemented to enable support for trait
interpolation to enable things like `{internal.kubernetes_group}` or even use
an external value from the auth provider.

Alternatively, we could introduce a new field `custom_kubernetes_groups`,
which would behave like the existing `kubernetes_groups` except it would be
optional and default to the Teleport cluster-admin role if not set. We can't
re-use the existing field as it would create a risk when upgrading role
version.

Finally, as we plan to implement Healthchecks soon, one of the check can be to
verify permissions and to surface a health issue if there is something wrong.

### Being authoritative

#### New Kubernetes Cluster Role / Binding

To make Teleport authoritative, we need a dedicated Cluster Role / Binding with
admin privileges.

We'll need to update the Helm chart, the `get-kubeconfig.sh` provisioning
script as well as the discovery service for ASK, EKS, and GCP.

> The new Cluster Role will be named `teleport-cluster-admin` and will have the

Author: Added a mention to refer to the Configuration section.

if they are set, the validation will reject the role.
If a role with version 9 attempts to set the `kubernetes_users` or
`kubernetes_groups` fields, or if an older role with those values set attempt
to update the version to 9, it will be rejected.

- https://github.com/gravitational/teleport/blob/22eb8c6645909a26d1493d01d291e222a87b35e6/api/types/role.go#L1979

#### Downgrade

When working with older agents not supporting the role version, we can't
downgrade the role, so we'll inject a wildcard deny to make sure there is no
unexpected access being granted.

This is handled as part of the grpcserver auth:

- https://github.com/gravitational/teleport/blob/22eb8c6645909a26d1493d01d291e222a87b35e6/lib/auth/grpcserver.go#L1950

### Configuration

To accommodate a scenario where a customer already has a Group with the preset
name, we need to add a new field in the kube_service configuration file to
define which admin group will be impersonated.

This can also be used for customer upgrading from older versions that don't
want to provision a new clusterrole/clusterrolebinding. The config value could
be set to `cluster-admin` or `system:masters` to remove the need to provision

> new resources.

Author: Done.

#### Helm Config

The Helm Chart will provide the ability to specify custom names for all

> resources, (clusterrole, clusterrolebinding) and for the Group.

Author: Done.

    networkpolicies, leases

This resulted in the following Teleport roles:

- `editor`, which maps to `cluster-admin` and can do everything.
- `access`, which maps to `edit` and can do most CRUD operations.
- `auditor`, which maps to `view` and can do read-only operations.

#### New Teleport preset roles

In order to be consistent and allow for seamless use of Kubernetes within
Teleport, we will introduce the following new preset roles:

- `kube-access`: Maps to the Kubernetes preset `edit`. Can CRUD most basic
  builtin resources and read some limited administrative resources.
  This would be used in conjunction with the `access` Teleport role to grant
  most non-administrative access to users.
- `kube-editor`: Maps to the Kubernetes preset `cluster-admin`. Provides full
  access to all Kubernetes resources, including CRDs.
  This would be used in conjunction with the `editor` Teleport role to grant
  full administrative access to users.
- `kube-auditor`: Maps the the Kubernetes preset `view`. Provides read-only
  access to some Kubernetes resources.
  This would be used in conjunction with the `auditor` Teleport role to grant

>   read-only access to users.

Reviewer 2: As this section is discussing consistency, I wonder if we should be consistent with the way we manage day-0 SSH access today.

We have the `access` role with a wildcard selector and we use the `logins` user traits to grant access to specific users. Database and windows access also work the same way with `database_users` and `windows_logins`.

Once the Kubernetes Bindings described in this RFD are created, we can provide access with the existing `access` preset and user traits, without creating new roles.

This approach would have the following advantages:
- consistency with the other protocols which don't have their dedicated presets
- reduce the probability of users shooting themselves in the foot with the awkward kube_group + kube_resoucres interactions
- leaner and reduces the number of Teleport presets

The guides would tell day 0 users to do:

```bash
tctl users add hugo --logins root --kubernetes-group teleport:preset:editor --roles=access
```

And they would have access by default.

Author: That's a good point. I updated the RFD to reflect this. Now we don't create any preset on the Teleport side but focus on the Kubernetes provisionning. Creating the bindings as described before and rely on the user trait to interpolate which K8s role will be used.

cc @klizhentas this simplifies even further but significantly changes the approved proposal, can you take another look?

#### Kubernetes Bindings

On the Kubernetes side, in order to be able to use those presets, we will need
to create the corresponding `ClusterRoleBindings` based on `kubernetes_groups`.

While there is an existing binding for `cluster-admin` via `system:masters`, it
is not available everywhere and there are no binding for `edit` nor `view`.
For consistency and reliability, we will create one binding for each preset
role.

- `kubernetes_groups` -> `ClusterRole`

> - `teleport:preset:access` -> `edit`
> - `teleport:preset:editor` -> `cluster-admin`
> - `teleport:preset:auditor` -> `view`

Reviewer 2: nit: I'd stick to kubernetes cluster role names in the CRB so it's clear what they bind to. Users might want to reuse them.

```suggestion
- `teleport:preset:edit` -> `edit`
- `teleport:preset:cluster-admin` -> `cluster-admin`
- `teleport:preset:view` -> `view`
```
